crates/runtime/src/datafusion/udf/bucket.rs

```diff
@@ -0,0 +1,357 @@
+    fn from(val: BucketError) -> Self {
+        DataFusionError::External(val.to_string().into())
+    }
+}
+
+#[derive(Debug)]
+pub struct Bucket {
+    signature: Signature,
+}
+
+impl Default for Bucket {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Bucket {
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            signature: Signature::any(2, Volatility::Immutable),
+        }
+    }
+}
+
+impl ScalarUDFImpl for Bucket {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn name(&self) -> &'static str {
+        "bucket"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, arg_types: &[DataType]) -> Result<DataType, DataFusionError> {
+        if arg_types.len() != 2 {
+            return Err(BucketError::InvalidArgumentCount {
+                count: arg_types.len(),
+            }
+            .into());
+        }
+        Ok(DataType::Int32)
+    }
+
+    fn invoke_with_args(&self, args: ScalarFunctionArgs) -> Result<ColumnarValue, DataFusionError> {
+        let args = args.args;
+        let num_args = args.len();
+        if num_args != 2 {
+            tracing::debug!("Invalid argument count: {num_args}");
+            return Err(BucketError::InvalidArgumentCount { count: args.len() }.into());
+        }
+
+        let num_buckets = match &args[0] {
+            ColumnarValue::Scalar(ScalarValue::Int64(Some(n))) => {
+                if *n <= 0 || *n > MAX_NUM_BUCKETS {
+                    return Err(BucketError::InvalidNumBuckets { num_buckets: *n }.into());
+                }
+                *n
+            }
+            arg => {
+                return Err(BucketError::InvalidFirstArgType { value: arg.clone() }.into());
+            }
+        };
+
+        tracing::trace!("Computing bucket with num_buckets: {num_buckets}");
+
+        match &args[1] {
+            ColumnarValue::Scalar(scalar) => {
+                let bucket = compute_bucket(scalar, num_buckets)?;
+                Ok(ColumnarValue::Scalar(bucket))
+            }
+            ColumnarValue::Array(array) => {
+                let buckets = compute_bucket_array(array, num_buckets)?;
+                Ok(ColumnarValue::Array(Arc::new(buckets)))
+            }
+        }
+    }
+}
+
+fn compute_bucket(scalar: &ScalarValue, num_buckets: i64) -> Result<ScalarValue, DataFusionError> {
+    if scalar.is_null() {
+        return Ok(ScalarValue::Int32(None));
+    }
+    let array = scalar.to_array()?;
+    let mut hashes = vec![0; 1];
+    let random_state = RandomState::new();
+    create_hashes(&[array], &random_state, &mut hashes)?;
+    Ok(ScalarValue::Int32(Some(
+        u64::try_from(num_buckets)
+            .and_then(|n| i32::try_from(hashes[0] % n))
+            .context(BucketLargerThanTypeSnafu)?,
+    )))
+}
+
+fn compute_bucket_array(array: &ArrayRef, num_buckets: i64) -> Result<Int32Array, DataFusionError> {
```

Comment:
This is going to be extremely slow because it processes each array item sequentially. It would be better to compute the hashes into an Arrow Array and then use one of the vectorized arrow compute kernels to compute the bucket positioning.

https://docs.rs/arrow/latest/arrow/compute/index.html

If there isn't a ready-made compute kernel we can use, then we should mimic the vectorized computation those kernels internally provide.